feat(otlp-sink): Add OTLP sink plugin for exporting spans to AWS X-Ray

huypham612 opened the pull request:
### Description
This PR introduces a new otlp sink plugin for Data Prepper that supports exporting OpenTelemetry spans to AWS X-Ray over HTTP using OTLP.

Key features:
* Sends ExportTraceServiceRequest messages over HTTP to X-Ray-compatible endpoints
* Supports gzip compression and AWS SigV4 signing
* Retries with exponential backoff and jitter on retryable HTTP errors
* Metrics integration (delivery latency, payload size, retries, errors, etc.)
* Batching support with max_events, max_batch_size, and flush_timeout thresholds
* Fully unit-tested at 100% coverage
* End-to-end and Performance tested with a local pipeline sending spans to a SigV4-signed OTLP endpoint (AWS X-Ray)

### Issues Resolved
Resolves #5663
 
### Check List
- [X] New functionality includes testing.
- [X] New functionality has a documentation issue. Please link to it in this PR.
  - [X] New functionality has javadoc added
- [X] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

KarstenSchnitter reviewed (commented):
Let's have a discussion in the minimal feature set. I am missing mTLS and gRPC support. I will do a more thorough review of this PR next week.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/buffer/OtlpSinkBuffer.java:
```diff
@@ -0,0 +1,209 @@
+        this.queue = new LinkedBlockingQueue<>(getQueueCapacity());
+        sinkMetrics.registerQueueGauges(queue);
+
+        this.workerThread = new Thread(this::run, "otlp-sink-buffer-thread");
```
Please make this a non-daemon thread. `setDaemon(false)`.

KarstenSchnitter replied:
I see another problem with this approach: If the worker thread dies for some reason, it is not restarted. But Data Prepper keeps running. Eventually the queue will fill up. Than Data Prepper will reject incoming data. But it will still not crash. This leads to a scenario, where Data Prepper keeps alive but is not doing the work it is supposed to do. I suggest to use an ExecutorService, that restarts the thread and ensure `this::run` gets rescheduled.

kkondaka replied:
We don't typically add another thread in sinks. Why have another thread and context switching etc? What's the advantage of this approach over what we already do in Opensearch/S3 sinks?

huypham612 replied:
Hi Krishna, the OTLP sink delivers spans to live trace ingestion endpoints like AWS X-Ray, which expect low-latency, streaming-style delivery. X-Ray enforces a 1MB limit per request, and spans that arrive too late may not appear in time for trace analysis.

To meet these constraints, the sink uses an internal queue and background thread. This lets us sub-batch large pipeline batches into smaller, size-bound requests and flush them on a timeout, even during low or bursty traffic. Without the thread, time-based flushing wouldn’t occur unless new data arrives. Without the queue, we couldn’t enforce payload-size limits independently of the pipeline. These components ensure timely, size-compliant, and predictable delivery for OTLP workloads.

huypham612 replied:
> I see another problem with this approach: If the worker thread dies for some reason, it is not restarted. But Data Prepper keeps running. Eventually the queue will fill up. Than Data Prepper will reject incoming data. But it will still not crash. This leads to a scenario, where Data Prepper keeps alive but is not doing the work it is supposed to do. I suggest to use an ExecutorService, that restarts the thread and ensure `this::run` gets rescheduled.

Thank you for the correction, I will make the update accordingly.

huypham612 replied:
> We don't typically add another thread in sinks. Why have another thread and context switching etc? What's the advantage of this approach over what we already do in Opensearch/S3 sinks?



Hi Krishna, the OTLP sink delivers spans to live trace ingestion endpoints like AWS X-Ray, which expect low-latency, streaming-style delivery. X-Ray enforces a 1MB limit per request, and spans that arrive too late may not appear in time for trace analysis.

To meet these constraints, the sink uses an internal queue and background thread. This lets us sub-batch large pipeline batches into smaller, size-bound requests and flush them on a timeout, even during low or bursty traffic. Without the thread, time-based flushing wouldn’t occur unless new data arrives. Without the queue, we couldn’t enforce payload-size limits independently of the pipeline. These components ensure timely, size-compliant, and predictable delivery for OTLP workloads.

dlvenable replied:
@huypham612 , We can use non-blocking request APIs instead. Then handle the close out (e.g. release) as the request response is received.

huypham612 replied:
> @huypham612 , We can use non-blocking request APIs instead. Then handle the close out (e.g. release) as the request response is received.

Thanks for the suggestion, I will switch the transport layer to use Armeria’s non-blocking request. This will remove the blocking retry loop and ThreadSleeper, since request retries and timeouts can now be handled asynchronously.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/buffer/OtlpSinkBuffer.java:
```diff
@@ -0,0 +1,209 @@
+     * Continues running as long as {@link #running} is true or the queue is not empty.
+     * Handles encoding failures, timeout-based flush, and final flush on shutdown.
+     */
+    private void run() {
```
Is there a particular reason you added a single thread for sending the data? The calls to `output` will run on a thread that you can use when the batch size is reached. This is how the other sinks work as well.

dlvenable replied:
@huypham612 , Data Prepper should be doing this already. We should double check if conditional routing somehow broke this. But it would call the `sink` for every processing iteration even if the events were empty. Other sinks expect this as well.

dlvenable commented on data-prepper-plugins/otlp-sink/build.gradle:
```diff
@@ -0,0 +1,121 @@
+    implementation 'org.hibernate.validator:hibernate-validator:8.0.1.Final'
+
+    // OkHttp
+    implementation 'com.squareup.okhttp3:okhttp:4.12.0'
```
We already have a few different HTTP clients in Data Prepper. We already use Armeria for the peer forwarding as a web client and we use it as our primary HTTP and gRPC server. Let's use Armeria for the client here too in order to avoid additional dependencies.

huypham612 replied:
> We already have a few different HTTP clients in Data Prepper. We already use Armeria for the peer forwarding as a web client and we use it as our primary HTTP and gRPC server. Let's use Armeria for the client here too in order to avoid additional dependencies.

Thanks for the feedback. We initially chose OkHttp for its simplicity and alignment with OTLP SDK usage, especially around retries, gzip, and SigV4 signing. That said, we’re happy to switch to Armeria if preferred for consistency, it looks like it can support all the same features with some minor adjustments.

dlvenable commented on data-prepper-plugins/otlp-sink/build.gradle:
```diff
@@ -0,0 +1,121 @@
+    implementation 'io.micrometer:micrometer-core'
+
+    // Unit Testing
+    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.9.1'
```
You don't need these two lines for JUnit.

dlvenable commented on data-prepper-plugins/otlp-sink/build.gradle:
```diff
@@ -0,0 +1,121 @@
+    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.9.1'
+    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.9.1'
+    testImplementation 'org.slf4j:slf4j-simple:2.0.7'
+    testImplementation 'org.mockito:mockito-core:5.10.0'
```
You don't need this.

dlvenable commented on data-prepper-plugins/otlp-sink/build.gradle:
```diff
@@ -0,0 +1,121 @@
+    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.9.1'
+    testImplementation 'org.slf4j:slf4j-simple:2.0.7'
+    testImplementation 'org.mockito:mockito-core:5.10.0'
+    testImplementation 'org.awaitility:awaitility:4.2.0'
```
You don't need this.

dlvenable commented on data-prepper-plugins/otlp-sink/build.gradle:
```diff
@@ -0,0 +1,121 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+plugins {
```
You don't need these four lines.

Our top-level project brings these in.

dlvenable commented on data-prepper-plugins/otlp-sink/build.gradle:
```diff
@@ -0,0 +1,121 @@
+    id 'jacoco'
+}
+
+jacoco {
```
Let's use the same tool version for JaCoCo from the parent. So remove these three lines.

dlvenable commented on data-prepper-plugins/otlp-sink/build.gradle:
```diff
@@ -0,0 +1,121 @@
+
+check.dependsOn integrationTest
+
+jacocoTestReport {
```
This part is configured in the root build.gradle.

dlvenable commented on data-prepper-plugins/otlp-sink/build.gradle:
```diff
@@ -0,0 +1,121 @@
+            limit {
+                counter = 'LINE'
+                value = 'COVEREDRATIO'
+                minimum = 0.95
```
Thanks! Can we get 100%?

huypham612 replied:
Yes, and agreed.

dlvenable commented on data-prepper-plugins/otlp-sink/build.gradle:
```diff
@@ -0,0 +1,121 @@
+    }
+}
+
+tasks.named('jacocoTestReport') {
```
You shouldn't need this. The root build.gradle is already outputting the HTML report.

huypham612 replied:
> You shouldn't need this. The root build.gradle is already outputting the HTML report.

Thanks, I will remove all the unnecessary dependencies.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/OtlpHttpSender.java:
```diff
@@ -0,0 +1,221 @@
+                    sinkMetrics.incrementPayloadGzipSize(compressedPayload.get().length);
+                    return;
+                }
+            } catch (final IOException ioException) {
```
Armeria has a utility that we can use for backoffs and retries. Can you make use of that?

`com.linecorp.armeria.client.retry.Backoff` - We use this in other places in the project.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/OtlpHttpSender.java:
```diff
@@ -0,0 +1,221 @@
+    public void send(@Nonnull final byte[] payload) throws IOException {
+        for (int attempt = 0; attempt <= maxRetries; attempt++) {
+            try {
+                final Optional<byte[]> compressedPayload = gzipCompressor.apply(payload);
```
We shouldn't re-compress for each failed attempt. This should be easier to handle using the Armeria Backoff tools as mentioned below.

KarstenSchnitter replied:
This also looks like a feature, that should be completely handled by the HttpClient. It would also be beneficial not to materialize the compressed data into a byte array. Couldn't you just use the OutputStream directly to create the request body?

huypham612 replied:
> We shouldn't re-compress for each failed attempt. This should be easier to handle using the Armeria Backoff tools as mentioned below.

Thanks, I will update as suggested.

huypham612 replied:
> This also looks like a feature, that should be completely handled by the HttpClient. It would also be beneficial not to materialize the compressed data into a byte array. Couldn't you just use the OutputStream directly to create the request body?

Hi Krasten, we compress to a byte[] before sending to support SigV4 signing, which requires the full payload to compute the signature; from my understanding, streaming compression isn’t compatible with that without a more complex signing implementation.

KarstenSchnitter replied:
Yeah, I have seen that part in the meantime.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/OtlpHttpSender.java:
```diff
@@ -0,0 +1,221 @@
+     * @throws IOException  when failed to send the payload.
+     */
+    public void send(@Nonnull final byte[] payload) throws IOException {
+        for (int attempt = 0; attempt <= maxRetries; attempt++) {
```
Please decouple the code for sending from backoffs a little more to reduce the method size and complexity.

kkondaka replied:
I am planning to move `https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/sqs-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/sqs/SqsSinkExecutor.java` to a strategy based common code later this month. Maybe you want to do part of that (or all of that) as part of your PR?

huypham612 replied:
> Please decouple the code for sending from backoffs a little more to reduce the method size and complexity.

Will update as suggested.

huypham612 replied:
> I am planning to move `https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/sqs-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/sqs/SqsSinkExecutor.java` to a strategy based common code later this month. Maybe you want to do part of that (or all of that) as part of your PR?

Hi Krishna, due to my on-call rotation last week and some upcoming work on the next milestone for our backend service, I’m afraid I won’t have the bandwidth to do a good job on the refactor at this time. I’d prefer to keep the scope focused for now, but happy to revisit or contribute to the shared strategy later if needed.

dlvenable reviewed (changes requested):
Thank you @huypham612 for this contribution!

KarstenSchnitter commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/buffer/OtlpSinkBuffer.java:
```diff
@@ -0,0 +1,209 @@
+     * @param sinkMetrics the metrics collector to use
+     */
+    public OtlpSinkBuffer(@Nonnull final OtlpSinkConfig config, @Nonnull final OtlpSinkMetrics sinkMetrics) {
+        this(config, sinkMetrics, null, null);
```
You could create the encoder and sender here to avoid the conditional in the private constructor.

huypham612 replied:
Nice! I will refactor. Thank you.

KarstenSchnitter commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/buffer/OtlpSinkBuffer.java:
```diff
@@ -0,0 +1,209 @@
+     */
+    public void add(final Record<Span> record) {
+        try {
+            queue.put(record); // block until space available; guaranteeing lossless delivery
```
Not sure, if it is ok to block the entire pipeline like this. Is it really guaranteeing lossless delivery, if Data Prepper crashes?

huypham612 replied:
> Not sure, if it is ok to block the entire pipeline like this. Is it really guaranteeing lossless delivery, if Data Prepper crashes?

You're right, this doesn't guarantee durability if Data Prepper crashes, since the queue is in-memory. The blocking behavior is intentional to provide in-process backpressure and avoid dropping spans during normal operation. This lets us preserve ordering and delivery guarantees while the system is healthy. We’ll clarify the comment to reflect that “lossless” applies to runtime conditions, not crash recovery.

KarstenSchnitter commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/configuration/ThresholdConfig.java:
```diff
@@ -0,0 +1,35 @@
+class ThresholdConfig {
+
+    @JsonProperty("max_events")
+    @Min(value = 1, message = "max_events must be at least 1")
```
Why not allow unbound, when the batch size is already configured?

huypham612 replied:
Good suggestion, I will update to allow unbounded max_events. Thanks.

KarstenSchnitter commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/GzipCompressor.java:
```diff
@@ -0,0 +1,66 @@
+     * @return Optional containing the compressed payload, or empty if compression failed.
+     */
+    @Override
+    public Optional<byte[]> apply(final byte[] payload) {
```
I am not sure about this signature. Why would I need an Optional here? The method could also return an empty array.

huypham612 replied:
Updated to use empty array. Thank you.

KarstenSchnitter commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/OtlpHttpSender.java:
```diff
@@ -0,0 +1,221 @@
+        }
+
+        final String responseBody = responseBytes != null ? new String(responseBytes, StandardCharsets.UTF_8) : "<no body>";
+        if (RETRYABLE_STATUS_CODES.contains(status)) {
```
Here you need to inspect the `retry-after` header to be considered for retry backoff. Otherwise you are not implementing the protocol completely.

KarstenSchnitter commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/OtlpHttpSender.java:
```diff
@@ -0,0 +1,221 @@
+
+        final String responseBody = responseBytes != null ? new String(responseBytes, StandardCharsets.UTF_8) : "<no body>";
+        if (RETRYABLE_STATUS_CODES.contains(status)) {
+            throw new IOException(String.format("Retryable error. Status: %d, Response: %s", status, responseBody));
```
Since you are about to catch this exception, please create a custom subclass. Also think about, whether you cannot change the control flow to work without exceptions for retryable errors.

huypham612 replied:
> Also think about, whether you cannot change the control flow to work without exceptions for retryable errors.

Good point, I’ll refactor the retry logic to use a boolean return value instead, so we can handle retries explicitly without relying on exceptions for control flow. Thank you!

KarstenSchnitter commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/OtlpHttpSender.java:
```diff
@@ -0,0 +1,221 @@
+                final String errorMessage = partial.getErrorMessage();
+                if (rejectedSpans > 0 || !errorMessage.isEmpty()) {
+                    LOG.warn("OTLP Partial Success: rejectedSpans={}, message={}", rejectedSpans, errorMessage);
+                    sinkMetrics.incrementRejectedSpansCount(rejectedSpans);
```
This should be a different counter than when some data transformation fails, e.g., in the GzipCompressor. I suggest to add a `FailedSpansCount` for cases, where spans were dropped but not rejected by the target.

huypham612 replied:
> This should be a different counter than when some data transformation fails, e.g., in the GzipCompressor. I suggest to add a `FailedSpansCount` for cases, where spans were dropped but not rejected by the target.

Thank you, we’ll add a separate FailedSpansCount metric for internal processing errors (e.g., Gzip or encoding failures) to differentiate them from spans explicitly rejected by the OTLP target endpoint.

KarstenSchnitter commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/ThreadSleeper.java:
```diff
@@ -0,0 +1,32 @@
+            Thread.sleep(millis);
+        } catch (final InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("Sleep interrupted", e);
```
Why do you not rethrow as an InterruptedException?

huypham612 replied:
> Why do you not rethrow as an InterruptedException?

The current wrapping is due to Consumer<Integer> not allowing checked exceptions. I can refactor this to use a custom Sleeper interface that lets us preserve and handle InterruptedException properly in the send() loop.

KarstenSchnitter replied:
Please do.

KarstenSchnitter commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/metrics/OtlpSinkMetrics.java:
```diff
@@ -0,0 +1,128 @@
+        pluginMetrics.counter("retriesCount").increment(1);
+    }
+
+    public void incrementRejectedSpansCount(final long count) {
```
This should only be used when parsing partial responses. Everywhere else a different counter should be used.

huypham612 replied:
> This should only be used when parsing partial responses. Everywhere else a different counter should be used.

Agreed, and will add a new metric.

KarstenSchnitter reviewed (commented):
I support @dlvenable suggestion to use another http client. I also suggest to be more distinctive on the error metrics, so we can distinguish message drops because of processing errors from those rejected by the target.

huypham612 commented:
Thank you for your feedback @dlvenable and @KarstenSchnitter, I will go through them and revise the code accordingly.

kkondaka commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/configuration/AwsAuthenticationConfig.java:
```diff
@@ -0,0 +1,37 @@
+ */
+@Getter
+@NoArgsConstructor
+class AwsAuthenticationConfig {
```
I have recently added `https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/aws-plugin-api/src/main/java/org/opensearch/dataprepper/aws/api/AwsConfig.java`. Can this be used instead of creating a new file?

huypham612 replied:
> I have recently added `https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/aws-plugin-api/src/main/java/org/opensearch/dataprepper/aws/api/AwsConfig.java`. Can this be used instead of creating a new file?

Yes, I will use this AwsConfig.java instead. Thank you!

huypham612 commented:
> I support @dlvenable suggestion to use another http client. I also suggest to be more distinctive on the error metrics, so we can distinguish message drops because of processing errors from those rejected by the target. Please also have a look at the `retry-after` header as described here: https://opentelemetry.io/docs/specs/otlp/#otlphttp-throttling

Thank you, I will address this comment and update the code to handle this header. Good catch.

huypham612 commented on data-prepper-plugins/otlp-sink/README.md:
```diff
@@ -0,0 +1,158 @@
+
+### Summary
+
+* Sustains ~3.5K TPS with ≤150ms p99 latency on t4g.large.
```
> Sorry, this is not what I would call "high-throughput" on the given infrastructure. Can you compare this to just the trace-source without the sink?

Thanks, that’s fair. I could remove the “high-throughput” phrasing to keep the language objective. The intent of the benchmark is to show relative improvements over the stock OTLP-HTTP and gRPC exporters. While we haven’t benchmarked the trace source in isolation, the current numbers reflect consistent performance gains and meet our latency and resource targets for the sink.

huypham612 commented on data-prepper-plugins/otlp-sink/README.md:
```diff
@@ -0,0 +1,158 @@
+
+## Retry Behavior
+
+- The sink uses an exponential backoff strategy for retryable HTTP status codes (e.g., 429, 502, 503, 504).
```
Thank you, I will address this in the revision.

huypham612 commented:
Hi @dlvenable, @KarstenSchnitter, and @kkondaka,

I've pushed the updated changes based on the feedback. Here’s a summary of the key updates:

    Migrated to Armeria HTTP client.

    Added failedSpansCount metric for improved observability.

    Introduced ExecutorService to manage the buffer queue and ensure resilience.

    Allowed unbounded max_events by supporting 0 as a valid value.

    Reused the shared AwsConfig class for consistency with other sinks.

    Re-ran performance tests and confirmed the implementation works as expected.

Note: I evaluated using the Retry-After header for dynamic backoff, but decided against it due to the following trade-offs:

    Armeria’s retry rule API only supports boolean predicates or fixed Backoff.

    Supporting Retry-After would require a custom Backoff implementation, which adds complexity with minimal benefit for most OTLP endpoints.

    The existing exponential backoff strategy already handles typical retry scenarios effectively.

Please take another look when you have time — thank you!

huypham612 commented on data-prepper-plugins/otlp-sink/README.md:
```diff
@@ -0,0 +1,158 @@
+
+The OTLP sink plugin sends span data using the OpenTelemetry Protocol (OTLP) format.
+The initial release supports exporting spans to AWS X-Ray. Future releases will support sending spans, metrics, and logs
+to any OTLP Protobuf-compatible endpoint.
```
I would love to continue contributing to the plugin and repo in the future.

kkondaka commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/buffer/OtlpSinkBuffer.java:
```diff
@@ -0,0 +1,213 @@
+
+                if (record != null) {
+                    try {
+                        final ResourceSpans resourceSpans = encoder.convertToResourceSpans(record.getData());
```
After the "record" is consumed, it must be "released" but only after it is successfully delivered. It looks like the entire "releasing" logic is missing.

kkondaka replied:
batch should hold resourceSpans and record's eventHandle. Something like 
```
batch.add(resourceSpans, record.getData().getEventHandle());
```
And release the eventHandle by doing `eventHandle.release(true)` if the eventual delivery to the external sink is successful. Replace true with false if fails.

huypham612 replied:
Hi @kkondaka,
For this first release, the plugin is stateless and does not integrate with DLQ or retry-aware sources. From my understanding, eventHandle.release(...) can be omitted for now for simplicity. We can add it in a follow-up if DLQ or retry support is introduced.

kkondaka replied:
Even if DLQ is not supported, `eventHandle.release()` cannot be skipped. Skipping release will result in events staying in memory for ever.

huypham612 replied:
Got it, I will update.

kkondaka commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/buffer/OtlpSinkBuffer.java:
```diff
@@ -0,0 +1,213 @@
+        final byte[] payload = request.toByteArray();
+
+        final int spans = batch.size();
+        sender.send(payload, spans);
```
This can't fail?

huypham612 replied:
Could you clarify what specific failure scenario you're concerned about?

The sender.send(...) method is asynchronous and internally handles retries for known OTLP retryable status codes (e.g., 429, 503) using Armeria’s retrying client. If all retries are exhausted or a non-retryable error occurs, it logs the failure and updates the appropriate metrics (rejectedSpansCount, failedSpansCount, errorsCount, etc.). Successful and partial responses are also tracked via recordsOut and rejectedSpans.

kkondaka replied:
Same as above. If send() fails  we need to do `release(false)` and if it succeeds, needs to do `release(true)` For that reason, it is important to know if send() succeeded or not

huypham612 replied:
Got it, will update the code.

kkondaka commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/buffer/OtlpSinkBuffer.java:
```diff
@@ -0,0 +1,213 @@
+            queue.put(record);
+        } catch (final InterruptedException e) {
+            Thread.currentThread().interrupt();
+            LOG.error("Interrupted while enqueuing span", e);
```
In case of failure, eventHandle should be released with "false" as the result.

kkondaka replied:
Why would it be interrupted? is it not possible to retry, if interrupted. Failing to add to buffer means data loss.

huypham612 replied:
The thread is typically interrupted during pipeline shutdown. In that case, retrying would delay or block shutdown. We restore the interrupt flag, log the drop, and increment error metrics. Yes, it results in data loss, but only during shutdown. In the future, we could route this to a DLQ if integrated.

kkondaka replied:
OK. If you are sure that interruption will only happen during shutdown, then it's ok.

huypham612 replied:
Thank you Krishna.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/configuration/OtlpSinkConfig.java:
```diff
@@ -0,0 +1,114 @@
+public class OtlpSinkConfig {
+
+    @Getter
+    @JsonProperty("endpoint")
```
We are generally using the term `host` or `url` in our projects. Do you think these would be accurate replacements? If there is a strong reason to prefer `endpoint` because it is the OTLP terminology, then we can use `endpoint`.

huypham612 replied:
We used endpoint here to stay aligned with the OTLP exporter spec and OpenTelemetry terminology (e.g., [OTLP Exporter config docs](https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/)).

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/SigV4Signer.java:
```diff
@@ -0,0 +1,101 @@
+ * before sending them to the AWS OTLP endpoint.
+ */
+class SigV4Signer {
+    private static final String SERVICE_NAME = "xray";
```
Rather than hard-code these, let's infer them from the endpoint itself.

The path should be easy to extract from the URL, right?

You could write something that is a little more extensible by creating a strategy for determining the service.

```
interface AwsServiceDetector {
  boolean matches(String endpoint);
  String getServiceName();
}
```

```
class XRayServiceDetector implements AwsServiceDetector {
  boolean matches(String endpoint) {
    return endpoint.contains("xray"); // Could be done more accurately as well by regex, but you get the idea
  }
  String getServiceName() {
    return "xray";
  }
}
```

Then in this class, take a `List<AwsServiceDetector>` in the constructor. Find the one that matches and get the service name.

huypham612 replied:
That makes sense for a more general plugin, but since this first release is specifically for X-Ray, I think we can keep it simple for now and hard-code the service name. If we expand to support other AWS services later, we can refactor with a detection strategy like this.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/SigV4Signer.java:
```diff
@@ -0,0 +1,101 @@
+                : URI.create(String.format("https://xray.%s.amazonaws.com%s", region.id(), OTLP_PATH));
+    }
+
+    private static AwsCredentialsProvider initCredentialsProvider(
```
Use the `AwsCredentialsSupplier` to get the credentials. This allows for a more consistent experience. It also allows for caching the credentials to avoid STS throttling.

https://github.com/opensearch-project/data-prepper/blob/81e4058652c44cddbf0eac123efb92473ea64fe0/data-prepper-plugins/aws-plugin-api/src/main/java/org/opensearch/dataprepper/aws/api/AwsCredentialsSupplier.java#L17-L31

huypham612 replied:
Sounds good, I will update.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/SigV4Signer.java:
```diff
@@ -0,0 +1,101 @@
+     * @param payload The OTLP Protobuf-encoded request body to be sent
+     * @return A signed {@link SdkHttpFullRequest} ready for transmission to the AWS OTLP endpoint
+     */
+    SdkHttpFullRequest signRequest(@Nonnull final byte[] payload) {
```
Per my other comment about being more dynamic, you probably also need to take the endpoint as a parameter.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/OtlpHttpSender.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    private HttpRequest buildHttpRequest(final byte[] compressedPayload) {
+        final SdkHttpFullRequest signedRequest = signer.signRequest(compressedPayload);
```
This should be based on a conditional.

```
final SdkHttpFullRequest request;
if(config.getAwsConfiguration() != null) {
  signedRequest = signer.signRequest(compressedPayload);
} else {
  signedRequest = // create unsigned request
}
```

Or you can add a method to the signer like `boolean doesEndpointRequireSigning(String endpoint)`.

Then you can have:

```
if(signer.doesEndpointRequireSigning(endpoint)) {
  final SdkHttpFullRequest signedRequest = signer.signRequest(compressedPayload);
  ...
```

huypham612 replied:
As aligned earlier, this first release is scoped to support only X-Ray, which always requires SigV4 signing. To keep things simple and focused, I’d prefer to defer conditional signing logic until we expand support beyond X-Ray.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/OtlpSink.java:
```diff
@@ -0,0 +1,89 @@
+/**
+ *  OTLP Sink Plugin for Data Prepper.
+ */
+@DataPrepperPlugin(
```
We can add the `@Experimental` annotation here if we are still developing the configuration and working toward completion.

huypham612 replied:
Will do, thanks David.

huypham612 commented on data-prepper-plugins/otlp-sink/README.md:
```diff
@@ -0,0 +1,157 @@
+
+- Currently, supports only trace data (spans). Support for metrics and logs will be added in future releases.
+- No support for DQL-based loss-less delivery in this release.
+- Only AWS X-Ray-compatible OTLP endpoints are currently supported (`https://xray.<region>.amazonaws.com/v1/traces`).
```
Yes, this is true for this release (Sigv4 is baked in).

huypham612 commented on data-prepper-plugins/otlp-sink/build.gradle:
```diff
@@ -0,0 +1,88 @@
+
+dependencies {
+    // AWS SDK
+    implementation 'software.amazon.awssdk:core:2.28.23'
```
Thanks, I will update.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/OtlpSink.java:
```diff
@@ -23,29 +25,34 @@
+    public OtlpSink(@Nonnull final AwsCredentialsSupplier awsCredentialsSupplier, @Nonnull final OtlpSinkConfig config, @Nonnull final PluginMetrics pluginMetrics, @Nonnull final PluginSetting pluginSetting) {
         super(pluginSetting);
 
+        config.validate();
```
Don't call this. Add an `@AssertTrue` on the `validate()` method instead. This will perform a Hibernate validation.

huypham612 replied:
Fixed. Thank you.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/configuration/OtlpSinkConfig.java:
```diff
@@ -97,18 +97,30 @@ public Region getAwsRegion() {
+     *
+     * @throws IllegalArgumentException if the AWS configuration is invalid
+     */
+    public void validate() {
```
Follow this pattern instead:

https://github.com/opensearch-project/data-prepper/blob/5dadde406be4038c6808795dcf4b6367e040f5b3/data-prepper-plugins/csv-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/csv/CsvInputCodecConfig.java#L72-L75

You can make it package protected.

huypham612 replied:
Updated. Thank you.

dlvenable commented on data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/http/OtlpHttpSender.java:
```diff
@@ -0,0 +1,236 @@
+        }
+
+        // Defensive copy to avoid ConcurrentModificationException
+        final List<Pair<ResourceSpans, EventHandle>> immutableBatch = List.copyOf(batch);
```
This is a shallow copy. So we are just copying the list. The underlying `ResourceSpans` and `EventHandle` objects are not copied.

dlvenable reviewed (approved):
Thank you @huypham612 for this great new feature!
